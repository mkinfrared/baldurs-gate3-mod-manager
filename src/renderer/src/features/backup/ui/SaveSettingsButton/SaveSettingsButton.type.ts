import { Game } from "@renderer/shared/types";

export type SaveSettingsButtonProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  game: Game;
};
