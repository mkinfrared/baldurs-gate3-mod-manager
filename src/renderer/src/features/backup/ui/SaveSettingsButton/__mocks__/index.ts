export { SaveSettingsButtonMock as SaveSettingsButton } from "./SaveSettingsButton.mock";
