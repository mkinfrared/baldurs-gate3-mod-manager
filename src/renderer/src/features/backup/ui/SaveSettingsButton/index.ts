export { SaveSettingsButton } from "./SaveSettingsButton";

export * from "./SaveSettingsButton.type";
