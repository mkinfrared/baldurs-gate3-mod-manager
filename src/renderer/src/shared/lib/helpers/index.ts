export { classNames } from "./classNames";

export { compose } from "./compose";

export { delay } from "./delay";

export { getNodeText } from "./getNodeText";

export { lazyImport } from "./lazyImport";

export { mergeRefs } from "./mergeRefs";

export { trpcClient, trpc } from "./trpc";

export * from "./typeGuards";
