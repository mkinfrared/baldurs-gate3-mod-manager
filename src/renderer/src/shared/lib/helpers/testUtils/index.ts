export { renderWithRouter, withQueryProvider } from "./testUtils";
