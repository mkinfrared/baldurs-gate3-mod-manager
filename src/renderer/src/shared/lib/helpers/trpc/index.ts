export { trpc, trpcClient } from "./trpc";
