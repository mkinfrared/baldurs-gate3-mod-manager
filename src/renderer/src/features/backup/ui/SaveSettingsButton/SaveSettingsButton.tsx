import { delay } from "@common/lib";
import { classNames, trpc } from "@renderer/shared/lib/helpers";
import { Button } from "@renderer/shared/ui";

import css from "./SaveSettingsButton.module.scss";
import { SaveSettingsButtonProps } from "./SaveSettingsButton.type";

const SaveSettingsButton = ({ className, game }: SaveSettingsButtonProps) => {
  const utils = trpc.useUtils();

  const { mutate, isLoading } = trpc.backup.createBackup.useMutation({
    onMutate: async (variables) => {
      await delay(850);

      return variables;
    },
    onSettled: () => {
      utils.backup.invalidate();
    },
  });

  return (
    <Button
      className={classNames(css.SaveSettingsButton, className)}
      data-testid="SaveSettingsButton"
      onClick={() => mutate(game)}
      loading={isLoading}
      disabled={isLoading}
    >
      Backup Current Settings
    </Button>
  );
};

export { SaveSettingsButton };
