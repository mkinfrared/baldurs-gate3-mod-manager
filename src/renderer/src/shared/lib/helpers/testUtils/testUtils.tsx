/* eslint-disable import/no-unused-modules */
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { ReactElement, ReactNode } from "react";
import { BrowserRouter, MemoryRouter } from "react-router-dom";

import { trpc, trpcClient } from "@renderer/shared/lib/helpers/trpc";

// https://testing-library.com/docs/example-react-router/
const renderWithRouter = (ui: ReactElement, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};

const queryClient = new QueryClient();

const withQueryProvider = (children: ReactNode) => (
  <trpc.Provider client={trpcClient} queryClient={queryClient}>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </trpc.Provider>
);

const withRouter = (children: ReactNode) => (
  <MemoryRouter>{children}</MemoryRouter>
);

export { renderWithRouter, withQueryProvider, withRouter };
