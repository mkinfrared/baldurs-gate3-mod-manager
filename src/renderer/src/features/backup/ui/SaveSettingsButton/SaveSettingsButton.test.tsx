import { render } from "@testing-library/react";

import { withQueryProvider } from "@renderer/shared/lib/helpers/testUtils";

import { SaveSettingsButton } from "./SaveSettingsButton";

describe("<SaveSettingsButton />", () => {
  const game = "BG3";
  const Component = <SaveSettingsButton game={game} />;

  it("should be defined", () => {
    expect(SaveSettingsButton).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(withQueryProvider(Component));

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(withQueryProvider(Component));
    const element = getByTestId("SaveSettingsButton");

    expect(element).toBeDefined();
  });
});
