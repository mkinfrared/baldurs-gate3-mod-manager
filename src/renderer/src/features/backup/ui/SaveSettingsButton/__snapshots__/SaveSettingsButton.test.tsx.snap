// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SaveSettingsButton /> > should match the snapshot 1`] = `
<div>
  <button
    class="Button SaveSettingsButton contained primary"
    data-testid="SaveSettingsButton"
    type="button"
  >
    <svg
      class="Spinner spinner"
    >
      bars.svg
    </svg>
    <span
      class="content"
    >
      Backup Current Settings
    </span>
  </button>
</div>
`;
